MAINT: address extraneous shape tuple checks in descriptor.c

tylerjereddy opened the pull request:
In looking for low-hanging fruit for fixes in the source `git grep 'TODO' | wc -l` suggests 120 source code comments where action might be helpful.

I noticed one source file where `TODO` comments suggested that code could simply be removed because it was effectively extraneous. Indeed, it seems that assuming that the object type is always consistent with `tuple` as the comments suggest has no effect on the results of the full test suite.

If these changes do increase vulnerability to some kind of issue(s), then at least that should show up in the peer review process of this PR, and I can add the missing tests; otherwise, perhaps it is indeed safe to remove those checks.

* C source comments flag two extraneous checks for
the object type of self->subarray->shape in
`descriptor.c`

* pertinent code has been adjusted to assume
that the objects are always tuples, as the comments
suggest; unit testing results are unchanged after
removal of alleged extraneous code

eric-wieser commented on numpy/core/src/multiarray/descriptor.c:
```diff
@@ -1938,15 +1938,8 @@ arraydescr_shape_get(PyArray_Descr *self)
-    /*TODO
-     * self->subarray->shape should always be a tuple,
-     * so this check should be unnecessary
-     */
```
I'd prefer to see these replaced with an `assert`

tylerjereddy replied:
Ok, I revised with assertions & adjusted title / commit message accordingly since we're actually preserving the tuple check, but at least removing the fallback return statement.

tylerjereddy commented:
A second set of revisions was also needed for ISO C90 standard compliance (apparently enforced by the CI compile) where type declarations are typically needed at the top of the block to avoid `ISO C90 forbids mixed declarations and code`. There's nothing wrong with what I did originally in C99 standard, but tried to refactor accordingly here.

eric-wieser commented:
We're tied to C90 due to MSVC on windows for python 2.7, I think. Once we drop 2.7 support, I think we can move to at least C99.

eric-wieser commented on numpy/core/src/multiarray/descriptor.c:
```diff
@@ -1938,33 +1939,23 @@ arraydescr_shape_get(PyArray_Descr *self)
-    return PyInt_FromLong(1);
+
+    assert(PyTuple_Check(self->subarray->shape));
+    ndim = PyTuple_Size(self->subarray->shape);
```
Good catch

eric-wieser commented on numpy/core/src/multiarray/descriptor.c:
```diff
@@ -1938,33 +1939,26 @@ arraydescr_shape_get(PyArray_Descr *self)
-    return Py_BuildValue("(O)", self->subarray->shape);
+    assert(PyTuple_Check(self->subarray->shape));
+    Py_INCREF(self->subarray->shape);
+    return (PyObject *)(self->subarray->shape);
```
While you're here, can you remove the pointless cast?

tylerjereddy committed:
MAINT: addressed extraneous shape tuple checks in descriptor.c

* C source comments flag two extraneous checks for
the object type of self->subarray->shape in
descriptor.c

* pertinent code has been adjusted to assert
the presence of the tuple object where
needed & remove a fallback return call along
with an extraneous casting

eric-wieser commented:
Thanks!
